lib/Vend/Tax/Avalara.pm

```diff
@@ -0,0 +1,1343 @@
+
+Handle case where the requested address involved in the lookup (internal or
+otherwise) is invalid.
+
+=cut
+
+sub invalid_address {
+    my $self = shift;
+    my $err = ::errmsg('%s %s is not a valid state/zip combination', $self->state, $self->zip);
+    ::logError('Request for tax returning 0 because %s', $err);
+    return 0;
+}
+
+=item C<_find_product_tax_code>
+
+Given a line item as an argument, find the corresponding Avalara product tax
+code. Using field B<product_tax_code_field> to look up first as a hash key in
+the line item, if exists, or fall back to C<item_field()> to find the field in
+the sku's source table from that same field name.
+
+=cut
+
+sub _find_product_tax_code {
+    my $self = shift;
+    my $item = shift;
+    my $k = $self->product_tax_code_field;
+
+    return $k && ($item->{$k} || Vend::Interpolate::item_field($item, $k)) || '';
+}
+
+=item C<_basic_args>
+
+Returns array mapping the standard Avalara attributes for any request to their
+corresponding Vend::Tax::Avalara attributes.
+
+=cut
+
+sub _basic_args {
+    my $self = shift;
+
+    return (
+        date => $self->transaction_date,
+        addresses => {
+            shipFrom => {
+                country => $self->nexus_country,
+                postalCode => $self->nexus_zip,
+                region => $self->nexus_state,
+                city => $self->nexus_city,
+                line1 => $self->nexus_address,
+            },
+            shipTo => {
+                country => $self->country,
+                postalCode => $self->zip,
+                region => $self->state,
+                city => $self->city,
+                line1 => $self->address1,
+            },
+        },
+        customerCode => $self->customer,
+    );
+}
+
+sub _handle_json_response {
+    my $self = shift;
+    my $resp = shift;
+
+    my $msg = $resp->decoded_content;
+
+    if ($resp->header('Content-Type') =~ /\bjson\b/) {
+        my $ref = $self->json->decode($msg);
+        _thin_booleans(\$ref);
+        push (@{ $self->http_response_stack }, { obj => $resp, doc => $ref }, );
+        return $ref;
+    }
+
+    push (@{ $self->http_response_stack }, { obj => $resp, content => $msg }, );
+
+    return { no_json => 1, content => $msg };
+}
+
+# Convert JSON::PP::Booleans to integer 1/0 recursively
+sub _thin_booleans {
+    my $ref = shift;
+
+    my $type = __PACKAGE__->_refd($$ref)
+        or return;
+
+    if ( $type eq 'JSON::PP::Boolean' ) {
+        $$ref = $$ref + 0;
+        return;
+    }
+
+    my $rtype = __PACKAGE__->_reftyped($$ref);
+    if ( $rtype eq 'HASH' ) {
+        _thin_booleans(\$_) for values %$$ref;
+    }
+    elsif ( $rtype eq 'ARRAY' ) {
+        _thin_booleans(\$_) for @$$ref;
+    }
+
+    return;
+}
+
+1;
+
+__END__
+
+=back
+
+=head1 AUTHOR
+
+Mark Johnson (mark@endpoint.com), End Point Corp.
+
+=head1 LICENSE AND COPYRIGHT
+
+Copyright (C) 2002-2023 Interchange Development Group
+
+Copyright (C) 1996-2002 Red Hat, Inc.
```

Comment:
All this Vend::Tax stuff is new, isn't it? Probably no Red Hat code in here.

Reply:
Likely not. I just put it in there to be consistent with the overall project, vs. piecemeal copyrighting.

Reply:
Ok, that makes sense. Nobody has to wonder later.